spec.html

```diff
@@ -25410,24 +25483,35 @@ <h1>Evaluate ( )</h1>
             <dd>a Cyclic Module Record _module_</dd>
 
             <dt>description</dt>
-            <dd>Evaluate transitions this module's [[Status]] from ~linked~ to ~evaluated~. If execution results in an exception, that exception is recorded in the [[EvaluationError]] field and rethrown by future invocations of Evaluate. (Most of the work is done by the auxiliary function InnerModuleEvaluation.)</dd>
+            <dd>Evaluate transitions this module's [[Status]] from ~linked~ to either ~evaluating-async~ or ~evaluated~. The Promise returned by Evaluate is allocated by the first invocation of Evaluate, and its capability is stored in the [[TopLevelCapability]] field. If execution results in an exception, the Promise is rejected. Future invocations of Evaluate return the same Promise. (Most of the work is done by the auxiliary function InnerModuleEvaluation.)</dd>
```

Comment:
Your wording sounds technically correct to me but may sound odd to someone not intimately familiar with the module loading semantics. To wit, "a component of a module graph" sounds more natural as "a component of the module graph", but you're highlighting that for an async module that's participating in module graphs, you get the promise of the first graph that invoked `Evaluate()`, right?

Reply:
Right.

Reply:
OK, given the discussion below I think my comment is not correct. Before I try again, let me try to capture the various conditions.

- The first time Evaluate is called on a module _m_ which is not part of any previously-imported graph, it will observe that _m_ is `~linked~` and will create a Promise for its evaluation, which it will store in _m_'s [[TopLevelCapability]].
- Subsequent calls to Evaluate for _m_, or to _any other module in its SCC_, will early-exit with the Promise created for _m_.
- Subsequent calls for another module _m2_ which is reachable from _m_ but not part of its SCC will see that _m2_ is `~evaluating-async~` or `~evaluated~`, but will still create a new Promise for _m2_'s evaluation. That will get stored in the [[CycleRoot]] for _m2_ (which may not be _m2_ itself), and will be returned on subsequent calls to Evaluate on _m2_ or any other module in its SCC.

Is that all correct?

Reply:
@bakkot yes that's exactly right, and as was noted in an earlier comment, the first importer effectively gets to define the cycle root.

Reply:
Given that, [[TopLevelCapability]] seems best conceptually described as a promise capability that exists in 1:1 with all SCCs in a graph. It is lazily created on first access, and the only way to access is via Evaluate().

Reply:
Great, thanks. So how's this:

```suggestion
            <dd>Evaluate transitions this module's [[Status]] from ~linked~ to either ~evaluating-async~ or ~evaluated~. The first time it is called on a module in a given strongly connected component, Evaluate creates and returns a Promise which resolves when the module has finished evaluating. This Promise is stored in the [[TopLevelCapability]] field of the root module for the component. Future invocations of Evaluate on any module in the component return the same Promise. (Most of the work is done by the auxiliary function InnerModuleEvaluation.)</dd>
```

Reply:
It is possible I am still confused, but if not, this wording runs into the confusion where the "root module" is not a property of the SCC but is programmatically determined by user code.

Reply:
That's true. I could change it to "the [[CycleRoot]] of the component", or possibly to "an arbitrarily chosen representative for the component".

Reply:
I prefer "the [[CycleRoot]] of the component", as that'll give readers a lead on what to grep for to figure out what that is.